### PR TITLE
fix(urls): middleware DI registrations survive every router-builder order

### LINKED ISSUE #4426: Design: auto-register SessionStore in DI singletons when SessionMiddleware is mounted
## Summary

`SessionMiddleware::new(config)` creates an internal `Arc<SessionStore>` but does **not** register it as a DI singleton. Any server function that uses `#[inject] session: SessionData` (or `#[inject] store: SessionStoreRef`) therefore fails at `ctx.get_singleton::<Arc<SessionStore>>()` with `DiError::NotFound`, which the `#[server_fn]` dispatcher maps to a generic 500. The two layers — middleware-driven session writes vs. DI-driven session reads — silently operate on different stores (or on no store at all, on the DI side).

This was surfaced by #4423 (anonymous POST to `create_question` returns 500). The fix shipped for #4423 (PR linked below) wires the store manually in `examples-tutorial-basis/src/config/urls.rs`, but the underlying API ergonomics issue remains.

## Motivation

Forgetting one line of DI wiring causes every `SessionData`-injecting handler to return an opaque 500 with no compile-time signal. This violates DESIGN_PHILOSOPHY:
- #5 *API ergonomics is paramount* — DI registration is invisible boilerplate that should not be hand-written for built-in framework primitives.
- #4 *Fail early* — currently fails at runtime with no diagnostic linking back to the middleware.
- #1 *Magic must be understandable* — auto-registration is acceptable if its mechanism is visible.

## Reproduction

`examples-twitter` and `examples-tutorial-basis` both call `SessionMiddleware::new(config)` and then `router.with_middleware(session_mw)` *without* `with_di_registrations(...)`. Every `#[inject] session: SessionData` server_fn under those routers will 500. The only known working wiring is the manual setup in `examples-twitter/src/apps/auth/tests/server_fn.rs:44-77`, which constructs `SingletonScope` directly.

### A. `Middleware` trait extension (auto-registration)

Add a default method:

```rust
// crates/reinhardt-http/src/middleware.rs
trait Middleware: Send + Sync {
    /* ... */
    fn di_registrations(&self) -> DiRegistrationList {
        DiRegistrationList::new()
    }
}
```

`SessionMiddleware` overrides it to register `Arc::clone(&self.store)`. `ServerRouter::with_middleware`  / `UnifiedRouter::with_middleware` extract and merge.

**Pros**: zero boilerplate at the call site; works for every middleware that owns DI-relevant state (cache, metrics, rate_limit, tracing, flatpages).
**Cons**: `reinhardt-http` would need to depend on (or re-export) `reinhardt-di::DiRegistrationList`. Default method on a public trait — backward-compatible but counts as a public API addition during RC. SP-6 / `rc-addition` review path required.

### B. Explicit builder helper

Expose a helper that returns the middleware + its registrations as a tuple, mirroring the existing `admin_routes_with_di(...)` shape:

```rust
let (session_mw, session_di) = SessionMiddleware::install(config);
router.with_middleware(session_mw).with_di_registrations(session_di);
```

**Pros**: no trait surface change; pattern already familiar from admin.
**Cons**: still two lines at the call site; easy to use `SessionMiddleware::new` accidentally and skip the helper.

### C. Builder method on existing types

`SessionMiddleware::di_registrations(&self) -> DiRegistrationList` plus convention to call it. Smallest diff but offers no guard rails.

## Open Questions

- Should other store-owning middleware (Cache, Metrics, Rate-limit, Tracing, Flatpages) opt into the same mechanism in this issue or follow-ups?
- Is the trait extension acceptable as an RC addition, or should it ship in `0.2.x`?

## Related

- Fixes (partial / application-layer): #4423
- PR fixing the symptom in `examples-tutorial-basis`: (link to be added on Draft PR creation)

Reproducibility surface: `crates/reinhardt-middleware/src/session/middleware.rs:71-89`, `crates/reinhardt-middleware/src/session/injectable.rs:38-88`, `crates/reinhardt-pages/macros/src/server_fn.rs:920-970`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### PR DESCRIPTION
**Pull request by kent8192**
## Summary

Follow-up fixes to PR #4435 (already merged). After merge, Codex stop-time review surfaced four distinct builder-call orderings that still silently dropped middleware-contributed DI registrations:

1. \`with_middleware(...)\` → \`with_di_context(...)\` — registrations went to the global deferred list, which startup skips when a user-supplied context exists. *(c9baaecb3)*
2. \`mount(child_with_pending)\` → \`with_di_context(ctx)\` on parent — child stayed without context, its pending list was later flushed to the global list and dropped. *(e1400c3cf)*
3. \`with_di_context(ctx)\` → \`mount(child_with_nested_grandchild)\` — \`mount\` drained the direct child but not nested grandchildren. *(8a6abdd65)*
4. \`group(routers_with_pending)\` was the last child-attachment entry point bypassing the recursive adopt walk entirely. *(95af00457)*

Root cause for all four: \`ServerRouter::with_middleware\` used to stash registrations into the global \`deferred_registration\` cell unconditionally; the merged PR fixed the obvious "context already attached" case but left subtree composition holes.

## Fix

- Added a router-owned \`pending_middleware_di: DiRegistrationList\` field. \`with_middleware\` stages into it when no context is attached.
- New private helper \`adopt_di_context_recursive(router, ctx)\` walks the subtree, draining every descendant's pending list into the context's \`SingletonScope\` and inheriting the context. Descendants that already own a different context are left alone.
- \`with_di_context\` / \`mount\` / \`mount_mut\` / \`group\` all route through this helper — single source of truth for context propagation.
- \`register_all_routes\` recursively flushes any still-staged entries to the global deferred list (covers the "no context anywhere" path that startup actually consumes).

## Type of Change

- [x] Bug fix (non-breaking change which fixes an issue)

## Motivation and Context

The bug was invisible to the existing \`SingletonScope::get\`-direct tests merged in #4435; the only handler-visible symptom was \`#[inject] session: SessionData\` returning \`DiError::NotFound\` at runtime under specific builder orderings. Codex stop-time review caught all four orderings in succession.

## How Was This Tested

Seven regression tests in \`crates/reinhardt-urls/src/routers/server_router/builder.rs\` (\`middleware_di_tests\` module), one per ordering plus the \"no context\" and nested-grandchild edge cases:

- \`with_middleware_before_with_di_context_applies_to_context\`
- \`with_middleware_after_with_di_context_applies_to_context\`
- \`with_middleware_without_context_flushes_to_global_on_register_all_routes\`
- \`child_pending_drains_into_parent_context_on_mount\`
- \`child_pending_drains_when_context_attached_after_mount\`
- \`nested_grandchild_pending_drains_into_parent_context_on_mount\`
- \`group_drains_grouped_router_pending_into_parent_context\`

All 7/7 pass. \`cargo nextest run -p reinhardt-urls -p reinhardt-middleware --all-features\` → 1729/1729 passed.

## Checklist

- [x] Code follows project style guidelines (\`cargo fmt\` / clippy clean on modified files)
- [x] Self-review of code completed
- [x] Code commented where necessary (each fix path references #4426 inline)
- [x] No new warnings introduced
- [x] Tests added that prove the fix is effective (7 regression tests, one per builder ordering)
- [x] All new and existing tests passing

## Labels to Apply

\`bug\` — addresses a runtime DI resolution failure under specific composition orders.

## Related Issues

Fix #4426
Refs #4435

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: crates/reinhardt-urls/src/routers/server_router/introspection.rs [hunk 1]
```diff
@@ -217,6 +224,20 @@ impl ServerRouter {
...
+	/// #4426.
+	fn flush_pending_middleware_di_recursive(router: &mut ServerRouter) {
+		if !router.pending_middleware_di.is_empty() && router.di_context.is_none() {
+			let pending = std::mem::take(&mut router.pending_middleware_di);
+			crate::routers::register_di_registrations(pending);
+		}
+		for child in router.children.iter_mut() {
+			Self::flush_pending_middleware_di_recursive(child);
+		}
+	}
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-urls/src/routers/server_router/introspection.rs:239]: Declined: the global-list flush relies on an invariant maintained by `mount`, `mount_mut`, and `group` — they call `adopt_di_context_recursive` which walks the subtree and sets `di_context` on every descendant that previously had none. By the time `register_all_routes` (the sole caller of `flush_pending_middleware_di_recursive`) runs, no descendant under an owning ancestor can have `di_context == None`, so a per-node check is sufficient and threading `inherited_ctx` through the recursion would be redundant. The doc comment is accurate given that invariant.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

File: crates/reinhardt-urls/src/routers/server_router.rs [hunk 2]
```diff
@@ -143,6 +143,15 @@ pub struct ServerRouter {
 	/// DI context
 	pub(crate) di_context: Option<Arc<InjectionContext>>,
 
+	/// Middleware-contributed DI singleton registrations that have been
+	/// harvested by [`with_middleware`] but not yet applied. Filled when
+	/// [`with_middleware`] runs before [`with_di_context`]; drained either by
+	/// a later [`with_di_context`] call (into that context's `SingletonScope`)
+	/// or, if no context is ever attached, by [`register_all_routes`] (into
+	/// the global deferred-registration list). This avoids both the silent
+	/// drop and the global-list leak described in #4426.
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-urls/src/routers/server_router.rs]: Addressed in 8d642a386: switched the four intra-doc links on `pending_middleware_di` to `[`Self::with_middleware`]` / `[`Self::with_di_context`]` / `[`Self::register_all_routes`]` so they resolve unambiguously from the struct-field context.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

File: crates/reinhardt-urls/src/routers/server_router/builder.rs [hunk 3]
```diff
@@ -314,9 +342,18 @@ impl ServerRouter {
...
+		// mirroring `with_di_context`. A non-recursive drain would leave
+		// nested grandchildren's staged registrations stranded; later
+		// `register_all_routes` would push them to the global list, which
+		// startup skips when the top router owns a context. See #4426.
+		if child.di_context.is_none()
+			&& let Some(parent_ctx) = self.di_context.as_ref()
+		{
+			Self::adopt_di_context_recursive(&mut child, parent_ctx);
+			child.di_context = Some(Arc::clone(parent_ctx));
 		}
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-urls/src/routers/server_router/builder.rs]: Addressed in 8d642a386: extracted the duplicated block into a private `ServerRouter::inherit_context_from_parent_if_any(parent, child)` helper, and replaced the inline blocks in `mount`, `mount_mut`, and `group` with calls to it. Existing middleware-DI tests in `builder.rs` still pass.

🤖 Generated with [Claude Code](https://claude.com/claude-code)
